### PR TITLE
Add SIMD implementations to some countWithinDistance functions

### PR DESCRIPTION
**Pull request by mvieth**
- Models are circle, plane, sphere
- One SSE and one AVX implementation each
- Benchmarks (1000 calls to countWithinDistance, compiled with -O3 -march=native):

| model   | current | SSE   | AVX   |
| ------- | ------- | ----- | ----- |
| plane   | 0.78s   | 0.42s | 0.38s |
| sphere  | 1.26s   | 0.45s | 0.37s |
| circle  | 1.08s   | 0.36s | 0.30s |

So the new functions are 2-3 times as fast.
The functions `selectWithinDistance()` and `getDistancesToModel()` could also be SIMD-ized in the future in a similar manner.
Other models that use normals are more difficult to SIMD-ize, mainly because of the `acos()` function used when computing angles. The `acos()` function might be approximated by other functions.

### CODE REVIEW

**Comment by mvieth** [comment]
Sorry, I forgot to remove some `#warning`s I added for debugging.

**Comment by kunaltyagi** [comment]
> The acos() function might be approximated by other functions.

I found [this](https://stackoverflow.com/a/46991254/2292816). No benchmarks, but something similar might help in other models. There's also [libsimdmath](https://webuser.fh-furtwangen.de/~dersch/) which has benchmarks on old models that might be helpful to inspire an implementation (I haven't looked at the license for both)

Thanks for the contribution 😄 

Also, as an aside, do you think using a library like [dimsum](https://github.com/google/dimsum), [xsimd](https://github.com/xtensor-stack/xsimd) or similar would help? I'm not recommending that, just asking.

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp]: If comments are required (for explanation), please insert:
```
// equivalent to the following:
```

Else remove unneeded comments

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp]: Maybe a unified macro for `__AVX__ && __AVX__2`? This function is AVX only, but the usage is when AVX2 is also present. What do you think? Same with SSE

**Comment by mvieth** [comment]
Those links are very nice finds! I had something like the first one in mind, with `sqrt` and some polynomial terms. There is also an implementation in Agner Fog's [library](https://github.com/vectorclass/version2/blob/master/vectormath_trig.h#L404). When choosing an approximating function, in the end it is all a matter of how small the error should be, while still making it as fast as possible.

> Also, as an aside, do you think using a library like dimsum, xsimd, ssemath or similar would help?

Using a library could help in matters of code readability and compatibility/portability. The downside would be having an additional dependency.

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp
Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp]: This function could be used in SIMD implementations of `selectWithinDistance()` and `getDistancesToModel()` in the future. I think those implementations would also be AVX only.

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp]: Yes, that's a bit nicer. Done.

**Comment by taketwo** [comment]
Hey, this is an interesting contribution, thanks. However, the fact that we are intensively sprinkling our code with `ifdef`ed SIMD instructions and don't provide any unit test coverage for them concerns me.

I'd like to avoid "contaminating" our code with `ifdef`ed SIMD instructions. Ideally, we would have a small helper functor/object that abstracts away the SIMD intrinsics. Something like `ComputeDistanceSSE()`, `ComputeDistanceAVX()`, `ComputeDistanceNoSIMD()`. Plus, `ComputeDistance()` that redirects to the most advanced implementation supported by the platform. In addition to making the "business" code easier to read, such abstraction will allow unit testing of different implementations.

It will take quite some work to modify PR in this way. Sadly, I'm really short on time currently and also would like to prioritize 1.10 release. So I won't be able to regularly iterate with reviews/discussions. @kunaltyagi do you have free cycles?

**Comment by kunaltyagi** [comment]
> do you have free cycles?

I do have the cycles. It also fits in with one of the RFC that I am working on, so that'll help. But it shouldn't take that much work. The implementations are already here.

1. Refactor them into functions
2. add tests (for 3 versions now)
3. Add central dispatch to route the current data flow
4. Update other models to follow the same thing, despite them not having SIMD implementations

> Plus, ComputeDistance() that redirects to the most advanced implementation supported by the platform

I agree. Central dispatch is better than `ifdef` because it allows individual testing, fine-grained choice at compile time using tags and less branches in code compilation. It also sounds like you're on-board the executor-tag proposal that I floated a while ago 😆

**Comment by mvieth** [comment]
Just to make sure I understand that correctly: the problem is that not all three versions are compiled and tested, so e.g. on a system where AVX is available, only that is tested, and a bug in the SSE implementation would never be found, right?

But that would also concern the `countWithinDistance` functions themselves.

**Comment by kunaltyagi** [comment]
> on a system where AVX is available, only that is tested, and a bug in the SSE implementation would never be found, right

Yes. If they were functions, they can be tested individually on a system with SSE + AVX. But in current form, only one of the 3 (SSE or AVX or non SIMD) can be tested

**Comment by mvieth** [comment]
I pushed a commit with some ideas. The 3 `countWithinDistance()` implementations are tested in different situations if they return the same value. The implementations are exposed as protected members to signalize that they are not intended for normal use. For the tests there is a helper class that inherits from the plane model to make the implementations available.
Since the SIMD distance functions (`dist4` and `dist8`) are used in the `countWithinDistance()` implementations, it is not really necessary to test them individually.

One question I don't yet have an answer to: what happens if SSE/AVX is not available?
- Should `countWithinDistanceSSE()` and `countWithinDistanceAVX()` be not defined? (preprocessor conditionals around the whole functions)
- Should they be defined and just do the computation the standard way (i.e. no SIMD, this is how it is currently in the commit)
- Should an exception be thrown if one of the functions is called

I'm looking forward to hear your opinion.

File: test/sample_consensus/test_sample_consensus_plane_models.cpp
Comment (mvieth) [reply test/sample_consensus/test_sample_consensus_plane_models.cpp]: Great idea, that looks much nicer!

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp
Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp]: Yes, you are right. That code should not be duplicated.

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp
Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp]: I wouldn't move them out of the class since they logically belong there. They also need access to the `input_`-member. Having an indices-parameter would IMO complicate things unnecessarily, and would also not really help with the "rest"-points in the SIMD implementations, since then they would have to construct a vector with the remaining indices (up to 3 or 7, respectively).

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp
Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp]: I'm not really familiar with the tag pattern. What are the advantages to the current state (with only preprocessor conditionals)? In your example, would `int func(int a)` correspond to `std::size_t countWithinDistance (const Eigen::VectorXf &model_coefficients, const double threshold)`, and the dispatching would happen inside that function? Then the user would not be able to choose the impl, would they? I'm not sure if that possibility is necessary or good, anyways. That transition penalty seems to be negligible, if it happens at all. At least I don't see where it could happen inside the pcl code.

File: test/sample_consensus/test_sample_consensus_plane_models.cpp
Comment (kunaltyagi) [inline test/sample_consensus/test_sample_consensus_plane_models.cpp]: Maybe put this in `ifdef` guard to ensure that tests compile even when AVX2 isn't available

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp
Comment (kunaltyagi) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp]: Yeah, that would be unweildy. Forgot that indices don't refer to the points themselves (like iterator or `std::reference_wrapper`)

File: sample_consensus/include/pcl/sample_consensus/sac_model_plane.h
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/sac_model_plane.h]: Maybe add an end option via an overload? That'll allow generic (seems to work) OpenMP implementation that can internally use the AVX and SSE functions.

[godbolt link](https://godbolt.org/z/cKLTvv)

Do note that I used tags, but you can skip that.

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp]: The code reuse in plane model looks good. Please replicate it elsewhere too. :+1:

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp
Comment (kunaltyagi) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp]: > What are the advantages to the current state (with only preprocessor conditionals)?

Tag pattern is a forward looking design (using C++ executor proposal). Right now in the examples, I'm not using the post, dispatch, and related functions but they can be added later on to conform to the standard.

Tag pattern also allows 
* more flexible usage and extensions
* separation of routing logic with implementation code

This doesn't benefit PCL much in the current design, but for instance, adding OMP will not require adding anything to existing functionality. I've added a godbolt link showing how the interface can be extended with both orthogonal (OpenMP) as well as complimentary features (AVX, SSE, etc.)

This obviously adds a complication and I'm fine with postponing this till I've gathered more consensus from other maintainers.

> the dispatching would happen inside that function?

The new link shows how. There are 2 user-facing functions, one with a sensible default and the other with a user input.

> That transition penalty seems to be negligible, if it happens at all

The transition is pretty common for hyper-threaded code since the kernel doesn't have insight into stuff like this. Moreover, AVX code brings down the clock speed by upto 15% and AVX512 by 50% (due to thermal constraints). This will not be seen in single threaded code, but multi-threaded code will be adversely affected if the threads are not all doing AVX at the same time. [See this link](https://www.anandtech.com/show/10158/the-intel-xeon-e5-v4-review/3) for a blog style overview.

File: sample_consensus/include/pcl/sample_consensus/sac_model_plane.h
Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/sac_model_plane.h]: That would be possible, but TBH I would leave that to the person who creates that OpenMP implementation, if that ever happens. Right now it would be just unused code.

File: test/sample_consensus/test_sample_consensus_plane_models.cpp
Comment (mvieth) [reply test/sample_consensus/test_sample_consensus_plane_models.cpp]: But the tests would compile even if SSE/AVX/AVX2 are not available? Then `countWithinDistanceSSE` and `countWithinDistanceAVX` would be equivalent to `countWithinDistanceStandard`, because the SIMD-part of those functions would be removed and only the call to `countWithinDistanceStandard` would remain.

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp
Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_sphere.hpp]: Ok, done.

File: sample_consensus/include/pcl/sample_consensus/sac_model_plane.h
Comment (kunaltyagi) [reply sample_consensus/include/pcl/sample_consensus/sac_model_plane.h]: > Right now it would be just unused code.

Sure

File: test/sample_consensus/test_sample_consensus_plane_models.cpp
Comment (kunaltyagi) [reply test/sample_consensus/test_sample_consensus_plane_models.cpp]: I thought using `countWithinDistanceAVX` will be a compile error if AVX is not available, but looking at the code that's not the case. That would be a better interface since it doesn't make false promises to the end-user. Function is available if functionality is available. What do you think about macros inside vs outside the functions?

Comment (mvieth) [reply test/sample_consensus/test_sample_consensus_plane_models.cpp]: Please read my comment from January 12 again, where I present 3 possibilities to handle that. I tend to agree to "no false promises".

Comment (kunaltyagi) [reply test/sample_consensus/test_sample_consensus_plane_models.cpp]: Sorry missed that.

**Comment by kunaltyagi** [comment]
> what happens if SSE/AVX is not available?

1. `Should countWithinDistanceSSE() and countWithinDistanceAVX() be not defined?`: 👍 
2. `no SIMD, silent fallback`: No, I don't like that the API makes a promise and fails silently
3. Should an exception be thrown if one of the functions is called: NO. This is detectable at compile time, so a compile time error is better than a run-time error.

Sorry for missing this for a week 😢

File: test/sample_consensus/test_sample_consensus_plane_models.cpp
Comment (mvieth) [reply test/sample_consensus/test_sample_consensus_plane_models.cpp]: Ok, I implemented option 1.

**Comment by kunaltyagi** [comment]
LGTM. @taketwo PTAL

The only thing I'd ask for is logging if the relevant tests ran, but
* it's not needed for the client since a passing test implies all is well
* developers can confirm using the macros (which they'll be using themselves)

**Comment by mvieth** [comment]
I added SIMD acos functions and SIMD implementations for normal plane. Let me know what you think.

**Comment by kunaltyagi** [comment]
Nice!!! Could you add some reference to the `acos` impl. And apart from cosmetic changes, this looks good.

**Comment by mvieth** [comment]
> Could you add some reference to the acos impl

Do you mean a source where I got the formula and the coefficients for the approximate acos function from? I got inspiration from some places, e.g. the links you and I posted above, but I came up with the formula in its current form, and I computed the coefficients with python by minimizing the average error.

**Comment by kunaltyagi** [comment]
> I computed the coefficients with python by minimizing the average error.

Since this is a custom coeff situation, maybe add this inside the function (not as documentation). A simple psudo-code (or better: python) to reproduce this for verification (and future) would be nice since different cos impl have different aim (speed, error, etc.).

Else I was just referring to a link/attribution to a trusted source where you got the formula. The tests passing indicate that the function is correct, but the details would prove it's valid everywhere (to a degree)

**Comment by mvieth** [comment]
> A simple psudo-code (or better: python)

Where's the difference? :wink:  :rofl: 

> Else I was just referring to a link/attribution to a trusted source where you got the formula

Like I said, I didn't find this formula somewhere, I came up with it myself. All the other implementations I have seen use fewer coefficients, meaning less accuracy.

File: common/include/pcl/common/impl/common.hpp
Comment (mvieth) [reply common/include/pcl/common/impl/common.hpp]: The point of the abs of the dot product is that it is the same as `angle=getAngle3D(...); angle=min(angle, pi-angle);` but the acos is only needed in the interval [0; 1]. Hm, I guess the name could be changed. My ideas would be `getMinAngle3DSSE` or `getAcuteAngle3DSSE`. What do you think?

Comment (kunaltyagi) [reply common/include/pcl/common/impl/common.hpp]: Acute is better than Min

Comment (mvieth) [reply common/include/pcl/common/impl/common.hpp]: I changed the name.

**Comment by kunaltyagi** [comment]
@taketwo How does this look? Apart from a few cosmetic changes, I think the PR looks to be in a good shape

**Comment by mvieth** [comment]
@kunaltyagi No worries!
I resolved the conflicts.

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp:113]: Please remove direct access to `points` since we're planning on some internal changes which will affect visibility of points in the future

**Review by kunaltyagi — DISMISSED** [review]

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp
Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp]: Done (at least in the parts where this PR makes changes)

File: sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp
Comment (kunaltyagi) [inline sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp:121]: Can we make a macro and undef it after these lines? Looks to be very easy place to make a error-by-copy in the future.

Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp:121]: A macro for what exactly? For `(*input_)[(*indices_)[INDEX]]`?

Comment (kunaltyagi) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp:121]: Actually, `(*this)[i]` should also work. See pcl_base.h L145

Comment (mvieth) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp:121]: But `SampleConsensusModelPlane`does not inherit from `PCLBase` right now. It might make sense to let `SampleConsensusModel` inherit from `PCLBase`.

Comment (kunaltyagi) [reply sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp:121]: Or implement it if inheritance isn't needed (or complicates matters)

**Review by kunaltyagi — COMMENTED** [review]
LGTM, apart from `(*input_)[(*indices_)[i+{0-7}]].x` repeated in short bursts of code

**Review by kunaltyagi — APPROVED** [review]

**Comment by kunaltyagi** [comment]
@mvieth Please test if you can push by merging this PR whenever you feel like 🎉

**Comment by mvieth** [comment]
What option do you usually use for PCL? Merge commit, squash and merge, or rebase and merge?

**Comment by kunaltyagi** [comment]
If multiple commits that should be 1, then squash. Else, after the author has rearranged/squashed properly, merge

**Pull request merged**
